File: Algorithms/Graphs/Python/KnuthMorrisPratt.py
def build_lps(pattern):
    lps = [0] * len(pattern)
    i = 0
    j = 1

    while j < len(pattern):
        if pattern[i] == pattern[j]:
            i += 1
            lps[j] = i
            j += 1
        else:
            if i != 0:
                i = lps[i - 1]
            else:
                lps[j] = 0
                j += 1

    return lps


def kmp_search(text, pattern):
    m = len(pattern)
    n = len(text)

    lps = build_lps(pattern)

    i = 0
    j = 0
    occurrences = []

    while i < n:
        if pattern[j] == text[i]:
            i += 1
            j += 1

            if j == m:
                occurrences.append(i - j)
                j = lps[j - 1]
        else:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1

    return occurrences


# Example usage:
text = "ABABDABACDABABCABAB"
pattern = "ABABCABAB"
result = kmp_search(text, pattern)
print("Pattern found at indices:", result)

File: Algorithms/Graphs/C/KnuthMorrisPratt.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void build_lps(char *pattern, int *lps)
{
    int len = 0;
    int i = 1;
    int m = strlen(pattern);

    lps[0] = 0;

    while (i < m)
    {
        if (pattern[i] == pattern[len])
        {
            len++;
            lps[i] = len;
            i++;
        }
        else
        {
            if (len != 0)
            {
                len = lps[len - 1];
            }
            else
            {
                lps[i] = 0;
                i++;
            }
        }
    }
}

int *kmp_search(char *text, char *pattern)
{
    int n = strlen(text);
    int m = strlen(pattern);

    int *lps = (int *)malloc(sizeof(int) * m);
    int *occurrences = (int *)malloc(sizeof(int) * n);

    build_lps(pattern, lps);

    int i = 0;
    int j = 0;
    int count = 0;

    while (i < n)
    {
        if (pattern[j] == text[i])
        {
            i++;
            j++;

            if (j == m)
            {
                occurrences[count] = i - j;
                count++;
                j = lps[j - 1];
            }
        }
        else
        {
            if (j != 0)
            {
                j = lps[j - 1];
            }
            else
            {
                i++;
            }
        }
    }

    occurrences = (int *)realloc(occurrences, sizeof(int) * count);
    occurrences[count] = -1;

    free(lps);
    return occurrences;
}

int main()
{
    char text[] = "ABABDABACDABABCABAB";
    char pattern[] = "ABABCABAB";

    int *result = kmp_search(text, pattern);

    printf("Pattern found at indices: ");
    for (int i = 0; result[i] != -1; i++)
    {
        printf("%d ", result[i]);
    }
    printf("\n");

    free(result);
    return 0;
}
